### PR TITLE
Add pending emails view for admins

### PR DESCRIPTION
**Pull request by Chessray**
Admins are now able to see pending emails.

Closes [AB#1176](https://dev.azure.com/ncia-anet/2aa083a5-af3d-44e1-8c7b-6e9e6b124d91/_workitems/edit/1176)

#### User changes
- None

#### Superuser changes
- None

#### Admin changes
- Additional menu option "Pending emails" displaying the pending emails with error messages where applicable.

#### System admin changes
- [ ] anet.yml or anet-dictionary.yml needs change
- [x] db needs migration
- [ ] documentation has changed
- [x] graphql schema has changed

### Checklist
  - [x] Described the user behavior in PR body
  - [x] Referenced/updated all related issues
  - [x] commits follow a `repo#issue: Title` title format and [these 7 rules](https://chris.beams.io/posts/git-commit/)
  - [x] commits have a [clean history](https://epage.github.io/dev/commits/), otherwise PR may be squash-merged
  - [x] Added and/or updated unit tests
  - [ ] Added and/or updated e2e tests
  - [x] Added and/or updated data migrations
  - [ ] Updated documentation
  - [x] Resolved all build errors and warnings
  - [ ] Opened debt issues for anything not resolved here

### CODE REVIEW

File: src/test/java/mil/dds/anet/test/TestData.java [hunk 1]
```diff
@@ -41,6 +47,11 @@ public static Position getTestAdvisor() {
     return b;
   }
 
+  public static AnetEmailInput createAnetEmailInput() {
+    return AnetEmailInput.builder().withToAddresses(List.of(FAKER.internet().emailAddress()))
+        .withComment(FAKER.text().text()).withCreatedAt(FAKER.timeAndDate().past(2, DAYS)).build();
+  }
```
Comment (gjvoosten) [inline src/test/java/mil/dds/anet/test/TestData.java:53]: Is it necessary to use random data here? Could just as well use fixed data, no?

File: src/test/java/mil/dds/anet/test/resources/AnetEmailResourceTest.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    // Search by author position
+    final AnetBeanList_Report results = withCredentials(adminUser,
+        t -> queryExecutor.reportList(getListFields("{ uuid }"), reportQuery));
+    final var reportUuid = results.getList().iterator().next().getUuid();
+    final var input = TestData.createAnetEmailInput();
+    final var mutationResult =
+        withCredentials(adminUser, t -> mutationExecutor.emailReport("", input, reportUuid));
+    assertThat(mutationResult).isEqualTo(1);
+
+    final var pendingEmails = withCredentials(adminUser, t -> queryExecutor.pendingEmails(""));
```
Comment (gjvoosten) [inline src/test/java/mil/dds/anet/test/resources/AnetEmailResourceTest.java:54]: You're not retrieving any fields in your query!

File: build.gradle [hunk 3]
```diff
@@ -279,6 +281,20 @@ test.environment = run.environment
...
+	}
+}
+
+eclipse {
+	classpath {
+		downloadSources = true
+		downloadJavadoc = true
+	}
+}
+
```
Comment (gjvoosten) [inline build.gradle:297]: Unrelated change, please revert.

File: build.gradle [hunk 4]
```diff
@@ -172,6 +172,8 @@ dependencies {
 	testImplementation 'org.junit.jupiter:junit-jupiter-params:5.10.3'
 	testRuntimeOnly 'org.junit.jupiter:junit-jupiter-engine:5.10.3'
 
+	testImplementation 'net.datafaker:datafaker:2.3.1'
```
Comment (gjvoosten) [inline build.gradle:175]: Probably not needed if you use fixed test data (see other comment).

File: client/src/pages/emails/Show.js [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    error,
+    pageProps: DEFAULT_PAGE_PROPS,
+    searchProps: DEFAULT_SEARCH_PROPS,
+    pageDispatchers
+  })
+  if (done) {
+    return result
+  }
+
+  console.log(data)
```
Comment (gjvoosten) [inline client/src/pages/emails/Show.js:42]: Please removing debugging remnants.

File: src/main/java/mil/dds/anet/AnetApplication.java [hunk 6]
```diff
@@ -413,12 +414,13 @@ public void run(AnetConfiguration configuration, Environment environment)
     final SubscriptionUpdateResource subscriptionUpdateResource =
         new SubscriptionUpdateResource(engine);
     final AttachmentResource attachmentResource = new AttachmentResource(engine);
+    final var emaiLResource = new AnetEmailResource(engine);
```
Comment (gjvoosten) [inline src/main/java/mil/dds/anet/AnetApplication.java:417]: Typo: emaiLResource → emailResource

File: src/main/java/mil/dds/anet/beans/AnetEmail.java [hunk 7]
```diff
@@ -20,6 +21,9 @@ public class AnetEmail {
   @GraphQLInputField
   private String comment;
 
+  @GraphQLQuery
```
Comment (gjvoosten) [inline src/main/java/mil/dds/anet/beans/AnetEmail.java:24]: Blank line above can be removed.

File: src/main/resources/migrations.xml [hunk 8]
```diff
@@ -5861,4 +5861,9 @@
 		</sql>
 	</changeSet>
 
+	<changeSet id="add-column-error-log-to-pending-emails" author="chessray">
+		<addColumn tableName="pendingEmails">
+			<column name="errorMessage" type="varchar(512)"/>
+		</addColumn>
+	</changeSet>
```
Comment (gjvoosten) [inline src/main/resources/migrations.xml:5868]: Please keep the blank line before the final closing `</databaseChangeLog>`.

File: client/src/pages/emails/Show.js [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+} from "components/Page"
+import moment from "moment"
+import React from "react"
+import { Table } from "react-bootstrap"
+import { connect } from "react-redux"
+import Settings from "settings"
+
+const GQL_GET_EMAILS = gql`
+  query {
+    allEmails: pendingEmails {
```
Comment (gjvoosten) [inline client/src/pages/emails/Show.js:18]: I would simply put `pendingEmails` here, and on line 54 use `data.pendingEmails.map(…)`.

File: client/src/models/AnetEmail.js [hunk 10]
```diff
@@ -0,0 +1,27 @@
+import Model from "components/Model"
+import * as yup from "yup"
+
+export default class AnetEmail extends Model {
```
Comment (gjvoosten) [inline client/src/models/AnetEmail.js:4]: This class is unused, it can be removed.

**Review by gjvoosten — CHANGES_REQUESTED** [review]

File: src/test/java/mil/dds/anet/test/resources/AnetEmailResourceTest.java (hunk 2 above)
Comment (Chessray) [reply src/test/java/mil/dds/anet/test/resources/AnetEmailResourceTest.java:54]: The assertion compares the retrieved email against the input (minus the fields that are assigned on persisting):

`    assertThat(pendingEmails).isNotEmpty()
        .anySatisfy(pendingEmail -> assertThat(pendingEmail).usingRecursiveComparison()
            .ignoringFields("__typename", "id", "createdAt").isEqualTo(input));`

Comment (gjvoosten) [reply src/test/java/mil/dds/anet/test/resources/AnetEmailResourceTest.java:55]: You should explicitly list the fields you want to retrieve in the query response, not use `""` (which is non-standard, and you only get away with it because the library is forgiving, and gets all fields incl. the private ones like `__type`). Look at other queries in the tests for some examples.

File: src/test/java/mil/dds/anet/test/resources/AnetEmailResourceTest.java [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+    final var input = TestData.createAnetEmailInput();
+    final var mutationResult =
+        withCredentials(adminUser, t -> mutationExecutor.emailReport("", input, reportUuid));
+    assertThat(mutationResult).isEqualTo(1);
+
+    final var pendingEmails = withCredentials(adminUser,
+        t -> queryExecutor.pendingEmails("{ id toAddresses createdAt comment errorMessage type }"));
+    assertThat(pendingEmails).hasSize(1)
+        .allSatisfy(pendingEmail -> assertThat(pendingEmail).usingRecursiveComparison()
+            .ignoringFields("__typename", "id", "createdAt", "type", "errorMessage")
```
Comment (gjvoosten) [inline src/test/java/mil/dds/anet/test/resources/AnetEmailResourceTest.java:59]: Why retrieve `"id", "createdAt", "type", "errorMessage"` if you're ignoring them anyway? You could just retrieve the ones you're interested in (`toAddresses comment`) and remove the call to `ignoringFields()`.

File: src/main/java/mil/dds/anet/threads/AnetEmailWorker.java [hunk 12]
```diff
@@ -94,7 +97,11 @@ protected void runInternal(Instant now, JobHistory jobHistory, Map<String, Objec
         }
         processedEmails.add(email.getId());
       } catch (Throwable t) {
-        logger.error("Error sending email", t);
+        logger.atError().setCause(t).setMessage("Error sending email #{}")
+            .addArgument(email.getId()).log();
```
Comment (gjvoosten) [inline src/main/java/mil/dds/anet/threads/AnetEmailWorker.java:101]: Why not simply: `logger.error("Error sending email {}", email.getId(), t);` ?

File: src/test/java/mil/dds/anet/test/resources/AnetEmailResourceTest.java (hunk 11 above)
Comment (Chessray) [reply src/test/java/mil/dds/anet/test/resources/AnetEmailResourceTest.java:59]: It validates the query itself. If the fields were not available, this would result in a GraphQL Error.

File: src/main/java/mil/dds/anet/resources/AnetEmailResource.java [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+public class AnetEmailResource {
+  private final EmailDao emailDao;
+
+  public AnetEmailResource(final AnetObjectEngine engine) {
+    this.emailDao = engine.getEmailDao();
+  }
+
+  @GraphQLQuery(name = "pendingEmails")
+  public List<AnetEmail> getPendingEmails(@GraphQLRootContext Map<String, Object> context) {
+    return emailDao.getAll();
```
Comment (gjvoosten) [inline src/main/java/mil/dds/anet/resources/AnetEmailResource.java:20]: This should be restricted to admins!

File: src/main/java/mil/dds/anet/threads/AnetEmailWorker.java (hunk 12 above)
Comment (Chessray) [reply src/main/java/mil/dds/anet/threads/AnetEmailWorker.java]: That wouldn't be the same thing. The only method with a particular `Throwable` argument is [this one](https://www.slf4j.org/apidocs-2.1.0/src-html/org/slf4j/Logger.html#line-814). Simply adding the `Throwable` as the final argument in a List treats it like any regular `Object`. https://www.slf4j.org/apidocs-2.1.0/src-html/org/slf4j/Logger.html#line-789

One thing I noticed though is that I'd really prefer for `Spotless` to align this call chain in a different way. Is there room for discussion on this?

File: src/test/java/mil/dds/anet/test/resources/AnetEmailResourceTest.java (hunk 11 above)
Comment (gjvoosten) [reply src/test/java/mil/dds/anet/test/resources/AnetEmailResourceTest.java:59]: To me it feels a bit pointless to check the GraphQL API in that way. In any case the `__typename` field can be removed from your list. And you could at least assert something on the `type`.

File: src/main/java/mil/dds/anet/threads/AnetEmailWorker.java (hunk 12 above)
Comment (gjvoosten) [reply src/main/java/mil/dds/anet/threads/AnetEmailWorker.java]: A Throwable as the last argument gets special treatment: https://www.slf4j.org/faq.html#paramException

File: src/test/java/mil/dds/anet/test/resources/AnetEmailResourceTest.java (hunk 11 above)
Comment (Chessray) [reply src/test/java/mil/dds/anet/test/resources/AnetEmailResourceTest.java]: Point taken on the `_typename`. There's an assertion for `type` two lines further down.

File: src/main/java/mil/dds/anet/resources/AnetEmailResource.java [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+public class AnetEmailResource {
+  private final EmailDao emailDao;
+
+  public AnetEmailResource(final AnetObjectEngine engine) {
+    this.emailDao = engine.getEmailDao();
+  }
+
+  @GraphQLQuery(name = "pendingEmails")
+  public List<AnetEmail> getPendingEmails(@GraphQLRootContext Map<String, Object> context) {
+    assertAdministrator(getUserFromContext(context));
```
Comment (gjvoosten) [inline src/main/java/mil/dds/anet/resources/AnetEmailResource.java:23]: To keep this in line with all other code, please import the classes `AuthUtils` and `DaoUtils`, and not the methods.

File: src/main/java/mil/dds/anet/threads/AnetEmailWorker.java [hunk 15]
```diff
@@ -94,7 +97,11 @@ protected void runInternal(Instant now, JobHistory jobHistory, Map<String, Objec
         }
         processedEmails.add(email.getId());
       } catch (Throwable t) {
-        logger.error("Error sending email", t);
+        logger.atError().setCause(t).setMessage("Error sending email #{}").addArgument(email::getId)
+            .log();
```
Comment (gjvoosten) [inline src/main/java/mil/dds/anet/threads/AnetEmailWorker.java:101]: Please simplify this to: `logger.error("Error sending email {}", email.getId(), t);`

File: src/test/java/mil/dds/anet/test/resources/AnetEmailResourceTest.java (hunk 11 above)
Comment (gjvoosten) [reply src/test/java/mil/dds/anet/test/resources/AnetEmailResourceTest.java:59]: Ah, I see, you're comparing an `AnetEmailInput` object to an `AnetEmail` object, and they have a different `__typename`…

Comment (gjvoosten) [reply src/test/java/mil/dds/anet/test/resources/AnetEmailResourceTest.java:59]: You could use:
```java
    assertThat(pendingEmails).hasSize(1)
        .allSatisfy(pendingEmail -> assertThat(pendingEmail).usingRecursiveComparison()
            .comparingOnlyFields("toAddresses", "comment").isEqualTo(input))
        […]
```
which is IMHO clearer (and simpler).

**Review by gjvoosten — APPROVED** [review]
